cmd/snap-bootstrap/cmd_scan_disk.go

```diff
@@ -0,0 +1,330 @@
+			}
+			if !same {
+				return nil
+			}
+		}
+	}
+
+	found := false
+	hasFallbackPartition := false
+	hasSeed := false
+	var seed_uuid string
+	hasBoot := false
+	var boot_uuid string
+	hasData := false
+	var data_uuid string
+	hasSave := false
+	var save_uuid string
+	for _, part := range partitions {
+		if !fallback {
+			if part.UUID == bootUUID {
+				found = true
+			}
+		}
+		if fallback && part.Name == fallbackPartition {
+			hasFallbackPartition = true
+		}
+		if part.Name == "ubuntu-seed" {
+			hasSeed = true
+			seed_uuid = part.UUID
+		} else if part.Name == "ubuntu-boot" {
+			hasBoot = true
+			boot_uuid = part.UUID
+		} else if part.Name == "ubuntu-data-enc" {
+			hasData = true
+			data_uuid = part.UUID
+		} else if part.Name == "ubuntu-data" {
+			hasData = true
+			data_uuid = part.UUID
+		} else if part.Name == "ubuntu-save-enc" {
+			hasSave = true
+			save_uuid = part.UUID
+		} else if part.Name == "ubuntu-save" {
+			hasSave = true
+			save_uuid = part.UUID
+		}
+	}
+
+	if (!fallback && found && (hasSeed || hasBoot)) || (fallback && hasFallbackPartition) {
+		fmt.Fprintf(output, "UBUNTU_DISK=1\n")
+		if hasSeed {
+			fmt.Fprintf(os.Stderr, "Detected partition for seed: %s\n", seed_uuid)
+			fmt.Fprintf(output, "UBUNTU_SEED_UUID=%s\n", seed_uuid)
+		}
+		if hasBoot {
+			fmt.Fprintf(os.Stderr, "Detected partition for boot: %s\n", boot_uuid)
+			fmt.Fprintf(output, "UBUNTU_BOOT_UUID=%s\n", boot_uuid)
+		}
+		if hasData {
+			fmt.Fprintf(os.Stderr, "Detected partition for data: %s\n", data_uuid)
+			fmt.Fprintf(output, "UBUNTU_DATA_UUID=%s\n", data_uuid)
+		}
+		if hasSave {
+			fmt.Fprintf(os.Stderr, "Detected partition for save: %s\n", save_uuid)
+			fmt.Fprintf(output, "UBUNTU_SAVE_UUID=%s\n", save_uuid)
+		}
+	}
+
+	return nil
+}
+
+func checkPartitionUUID(output io.Writer, suffix string, partUUID string) {
+	varname := fmt.Sprintf("UBUNTU_%s_UUID", suffix)
+	expectedUUID := osGetenv(varname)
+	if len(expectedUUID) > 0 && expectedUUID == partUUID {
+		fmt.Fprintf(os.Stderr, "Detected partition as %s\n", suffix)
+		fmt.Fprintf(output, "UBUNTU_%s=1\n", suffix)
+	}
+}
+
+func scanPartitionNode(output io.Writer, node string) error {
+	fmt.Fprintf(os.Stderr, "Scanning partition %s\n", node)
+
+	probe, err := blkid.NewProbeFromFilename(node)
+	if err != nil {
+		return err
+	}
+	defer probe.Close()
+
+	probe.EnablePartitions(true)
+	probe.SetPartitionsFlags(blkid.BLKID_PARTS_ENTRY_DETAILS)
+	probe.EnableSuperblocks(true)
+
+	if err := probe.DoSafeprobe(); err != nil {
+		return fmt.Errorf("Cannot probe partition %s: %s\n", node, err)
+	}
+
+	partUUID, err := probe.LookupValue("PART_ENTRY_UUID")
+	if err != nil {
+		return fmt.Errorf("Cannot get uuid for partition: %s\n", err)
+	}
+
+	for _, suffix := range []string{"SEED", "BOOT", "DATA", "SAVE"} {
+		checkPartitionUUID(output, suffix, partUUID)
+	}
+
+	return nil
+}
+
+func ScanDisk(output io.Writer) error {
```

Comment:
I think it is worth adding here a more detailed explanation of how this works: disk is first scanned, then env is exported, there is another invocation of scan for partitions that has can read this and this var from the env that were created by the disk scan, etc.